// home/index.js
const querystring = require('querystring')
const debug = require('debug')('getleverage')
const xhr = require('xhr')
const ClassList = require('class-list')
const domEvent = require('dom-events')
const Ladda = require('ladda')

const emailRE = /^[-a-z0-9~!$%^&*_=+}{'?]+(\.[-a-z0-9~!$%^&*_=+}{'?]+)*@([a-z0-9_][-a-z0-9_]*(\.[-a-z0-9_]+)*\.(\w{2,})|([0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}))(:[0-9]{1,5})?$/i

module.exports = function () {
  const buttons = document.querySelectorAll('button.signup')
  const errorState = document.querySelector('.error')

  const quoteElements = document.querySelectorAll('[name="quotes"]')
  const quotes = [].slice.call(quoteElements)
  let quote = quotes[0]
  let quoteIndex = 0

  setInterval(() => {
    quote = quotes[quoteIndex++] || quotes[quoteIndex = 0, quoteIndex++]
    quote.checked = true
  }, 3200, true)

  function showError (msg) {
    errorState.textContent = msg
    ClassList(errorState).add('show')
  }

  function submit (ladda, input, cb) {
    if (input && input.value) {
      input.value = input.value.toLowerCase().trim()
    }

    if (!input || !input.value || !emailRE.test(input.value)) {
      return ClassList(input).add('invalid')
    } else if (input) {
      ClassList(input).remove('invalid')
    }

    const payload = { email: input.value }

    const opts = {
      url: `${window.endpoint}/subscribe/email`,
      method: 'POST',
      json: payload
    }

    ladda.start()

    xhr(opts, function (err, res, data) {
      debug('response', res.statusCode, data)
      ladda.stop()

      if (err) {
        return showError(err.messasge)
      } else if (res.statusCode !== 200) {
        return showError([
          'We are currently experiencing an unknown error.',
          'We restarted our server though, can you please try again?'
        ].join(' '))
      }

      cb(input, payload)
    })
  }

  const $form = document.querySelector('form')

  domEvent.on($form, 'submit', function (event) {
    event.preventDefault()
  })

  function getCallback(btn) {
    const cl = ClassList(btn)

    if (cl.contains('service')) {
      return (input, payload) => {
        const loc = [
          window.location.href,
          'subscribe/?',
          querystring.stringify(payload)
        ].join('')

        window.location = loc
      }
    }
    else {
      return (input, payload) => {
        const success = `
          <div class="newsletter-success">
            <span>Added to mailing list!</span>
            <svg class="checkmark" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 52 52">
              <circle class="checkmark__circle" cx="26" cy="26" r="25" fill="none"/>
              <path class="checkmark__check" fill="none" d="M14.1 27.2l7.1 7.2 16.7-16.8"/>
            </svg>
          </div>
        `

        const parent = input.parentElement
        parent.innerHTML = success
      }
    }
  }

  Array.from(buttons).map(function (el) {
    const ladda = Ladda.create(el)
    const cb = getCallback(el)
    let input = el.previousSibling

    while (input && input.tagName !== 'INPUT') {
      input = input.previousSibling
    }

    domEvent.on(input, 'keydown', function (event) {
      if (event.keyCode === 13) submit(ladda, input, cb)
    })

    domEvent.on(el, 'click', function (event) {
      event.preventDefault()
      submit(ladda, input, cb)
    })
  })
}
